### ISSUE TITLE
Multi GPU  training for high resolution U-Net of the cascade 

### ISSUE THREAD

**Issue by WonderSimiliar** [MIC-DKFZ/nnUNet#292]
Hi, Fabian,

Could I help you to implement the "**Multi GPU  training for high resolution U-Net of the cascade**" ?

I have noticed that "_Distributed training currently only works for the basic trainers (2D, 3D full resolution and 3D low resolution) and not for the second, high resolution U-Net of the cascade. The reason for this is that distributed training requires some changes to the network and loss function, requiring a new nnUNet trainer class. This is, as of now, simply not implemented for the cascade, but may be added in the future._" in Readme.md

Could you please tell me more details of tips which can  solve the above problem?
I am eager to use _**Multi GPU  training for high resolution U-Net of the cascade**_.

Thanks for your sharing of nnUnet.

**Comment by FabianIsensee** [comment]
Hi there,
that you very much for the offer! I am not sure if there is a real benefit in doing that:
- using multi-GPU to speed up the training is not a very good use of GPU resources. Scaling is never perfect. It is better to run multiple experiments (all folds, multiple configurations) in parallel
- if you have a proper environment, your epoch times should be well below 200s for the 3d configurations and much faster for 2d. Training times are not that long, really. [proper environment means 1) enough CPU to keep the GPU busy, 2) Volta architecture (V100, Titan V) OR Turing (RTX 2080 ti, Titan RTX) using pytorch compiled with cuDNN 8.0.2]
- using multi-GPU to increase the batch size only has benefits on much larger datasets (which are often not available in our domain)
- using multi-GPU to enable larger patch sizes has not resulted in an increase in segmentation accuracy so far. Working on it.
- nnU-Nets multi-GPU inference is not well implemented as of now

I am not saying that it should not be done, but there is a lot more that has to be considered than just implementing that missing trainer class. I think it makes most sense I I do it once I have the time because it might take you a while to get used to everything in nnU-Net and it may actually be longer for me to explain everything than implement it myself.
Best,
Fabian

**Comment by WonderSimiliar** [comment]
Hi, Fabian,

There is a huge need for the Multi GPU training for high resolution U-Net of the cascade.

- I have a very large task to validate nnUnet. Each image's size is 300*2000*2000, and there is 1000+ images. so the whole data is too large.

- For each epoch(batch size=2), the fullRes nnUnet would take 6000 seconds (V100,32GB). And the whole train would take more than 1 week, which is not acceptable. For normal task, the V100 takes only 200-300s per epoch.

- fullRes nnunet in not good at 300*2000*2000's segmentation, while cascade nnunet is good at it.

So I am interested in Multi GPU training for high resolution U-Net of the cascade, and hope to implement it.
Thanks for your time！

**Comment by FabianIsensee** [comment]
Hi,
when you encounter such a long training time it makes sense to first identify the bottleneck. It is not the GPU, but the CPU. And I believe I know where it comes from. This will not be an easy fix but something I should be able to do next week.
Best,
Fabian

**Comment by WonderSimiliar** [comment]
Wow, it is fansantistic! Thanks!

**Comment by FabianIsensee** [comment]
I have implemented the change and I see a big improvement in CPU and I/O Performance. I am currently evaluating it internally and if everything works as expected I will push it to master.
Best,
Fabian

**Comment by WonderSimiliar** [comment]
> I have implemented the change and I see a big improvement in CPU and I/O Performance. I am currently evaluating it internally and if everything works as expected I will push it to master.
> Best,
> Fabian

Thanks a lot, Fabian!

**Comment by FabianIsensee** [comment]
Please `pip install --upgrade nnunet` or use the most recent master. This should be much better now (you will need to rerun preprocessing!)

**Comment by WonderSimiliar** [comment]
> Please `pip install --upgrade nnunet` or use the most recent master. This should be much better now (you will need to rerun preprocessing!)

Thanks, Fabian!

**Comment by FabianIsensee** [comment]
is it faster now?

**Comment by WonderSimiliar** [comment]
> is it faster now?

It is faster now！Fabian，thanks  a lot！

**Comment by FabianIsensee** [comment]
How fast is it per epoch?
